# Python express course for hardcores ;) - code instead of text :D

## Help

In [1]:
x = 1

In [2]:
dir(x)

['__abs__',
 '__add__',
 '__and__',
 '__bool__',
 '__ceil__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__float__',
 '__floor__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__index__',
 '__init__',
 '__init_subclass__',
 '__int__',
 '__invert__',
 '__le__',
 '__lshift__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__or__',
 '__pos__',
 '__pow__',
 '__radd__',
 '__rand__',
 '__rdivmod__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rfloordiv__',
 '__rlshift__',
 '__rmod__',
 '__rmul__',
 '__ror__',
 '__round__',
 '__rpow__',
 '__rrshift__',
 '__rshift__',
 '__rsub__',
 '__rtruediv__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__truediv__',
 '__trunc__',
 '__xor__',
 'as_integer_ratio',
 'bit_length',
 'conjugate',
 'denominator',
 'from_bytes',
 'imag',
 'numerator',
 'real',
 'to_bytes']

In [4]:
help(x)

Help on int object:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __ceil_

## Lists

In [5]:
x = [1,2,3]

In [6]:
x

[1, 2, 3]

In [7]:
x[1]

2

In [8]:
x[4]

IndexError: list index out of range

In [9]:
len(x)

3

In [10]:
x[-1]

3

In [11]:
x[:-1]

[1, 2]

In [12]:
x = [1,2,3,4,5,6,7,8,9]

In [13]:
x[::2]

[1, 3, 5, 7, 9]

In [14]:
x[::3]

[1, 4, 7]

In [15]:
x.append(10)

In [16]:
x

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [17]:
x.pop()

10

In [18]:
x.pop(0)

1

## List comporehension

In [19]:
[x for x in [1,2,3,4,5,6,7,8,9]]

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [20]:
[x for x in [1,2,3,4,5,6,7,8,9] if x % 2 == 0]

[2, 4, 6, 8]

In [21]:
[x*x for x in [1,2,3,4,5,6,7,8,9] if x % 2 == 0]

[4, 16, 36, 64]

## Dicts

In [22]:
x = {'a':1, 'b':2, 'c':3, 'd':4, 'e':5}

In [24]:
x['a']

1

In [26]:
x['f']=6

In [27]:
x

{'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5, 'f': 6}

In [28]:
x.keys()

dict_keys(['a', 'b', 'c', 'd', 'e', 'f'])

In [29]:
for key in sorted(x):
    print('{} -> {}'.format(key, x[key]))

a -> 1
b -> 2
c -> 3
d -> 4
e -> 5
f -> 6


Not existing element

In [31]:
element = x['unexisting']

KeyError: 'unexisting'

In [32]:
if 'unexisting' in x:
    element = x['unexisting']
else:
    element = 0

In [33]:
element

0

In [34]:
element = x['unexisting'] if 'unexisting' in x else 0

In [35]:
element

0

## Touples (read only lists)

In [36]:
x = (1,2,3,4,5,6,7,8,9)

In [38]:
x[0]=0

TypeError: 'tuple' object does not support item assignment

## Files

In [39]:
f = open('file.txt','w')

In [40]:
f.write('Line1\n')

6

In [41]:
f.write('Line2\n')

6

In [42]:
f.write('Line3 with additional text\n')

27

In [43]:
f.close()

In [44]:
f = open('file.txt', 'r')

In [45]:
text = f.read()

In [46]:
text

'Line1\nLine2\nLine3 with additional text\n'

In [52]:
f.close()

Binary files

In [53]:
f = open('file.txt','rb')

In [54]:
data = f.read()

In [55]:
data

b'Line1\r\nLine2\r\nLine3 with additional text\r\n'

In [56]:
f.close()

## Sets

In [57]:
x = {1,2,3,4,5,6,7}

In [58]:
x = set([1,2,3,4,5,6,7])

In [59]:
x

{1, 2, 3, 4, 5, 6, 7}

In [60]:
x = set([1,1,1,1,1,1,1,1,2,3,4,5,6,7])

In [61]:
x

{1, 2, 3, 4, 5, 6, 7}

In [66]:
x = {1,1,1,2,3}

In [67]:
x

{1, 2, 3}

Set from few lists:

In [72]:
x = [1,2,2,2,3,3,3]

In [73]:
y = [1,2,2,2,3]

In [74]:
z = [1,2,3,3,3]

In [77]:
set(x + y + z)

{1, 2, 3}

In [79]:
{1,2,3}&{1,2}

{1, 2}

In [80]:
{1,2,3}|{1,2}

{1, 2, 3}

In [81]:
{1,2,3}-{1,2}

{3}

## Function

In [295]:
def function():
    return ''

In [296]:
function()

''

In [322]:
def function(a,b=1): # b has default value equal to 1
    return 'a:{} b:{}'.format(a,b)

In [323]:
function('1')

'a:1 b:1'

In [300]:
function(a=10,b=20)

'a:10 b:20'

In [301]:
function(10,b=20)

'a:10 b:20'

In [324]:
def function(a,b=1,c:str='test'): # c has defined type as string
    print('type a:{} type b:{} type c:{}'.format(type(a), type(b), type(c)))
    return 'a:{} b:{} c:{}'.format(a,b,c)

In [325]:
function(1)

type a:<class 'int'> type b:<class 'int'> type c:<class 'str'>


'a:1 b:1 c:test'

In [312]:
function(1,c=12) # but you can pass here also int ;)

type a:<class 'int'> type b:<class 'int'> type c:<class 'int'>


'a:1 b:1 c:12'

In [326]:
def function(a,b=1,c:str='test') -> int: # you can even define what should be returned
    return 'a:{} b:{} c:{}'.format(a,b,c)

In [327]:
type(function(1)) # but who cares?

str

In [335]:
def function(a:"This is real important number :P", b=1, c:str='test') -> str: # some custom string as type for a ;)
    return 'a:{} b:{} c:{}'.format(a,b,c)

In [318]:
function(1)

'a:1 b:1 c:test'

In [333]:
def function(a, *args, **kwargs):
    print(args)
    print(kwargs)

In [334]:
function(1,2,3,something='string')

(2, 3)
{'something': 'string'}


Since python 3.8 function definition can be even more crazy (https://docs.python.org/3/whatsnew/3.8.html)

In [329]:
def function(a, /, b, *, c:str='test'): # a cannot be passed as keyword, c must be keyword
    return 'a:{} b:{} c:{}'.format(a,b,c)

In [330]:
function(a=1)

TypeError: function() got some positional-only arguments passed as keyword arguments: 'a'

In [331]:
function(1,2,3)

TypeError: function() takes 2 positional arguments but 3 were given

In [332]:
function(1,2,c=3)

'a:1 b:2 c:3'

## Class

In [99]:
class Something:
    
    def __init__(self, a, b=3):
        self.a = a
        self.b = b
        self.c = None
        
    def returnA(self):
        return self.a
    
    def returnB(self):
        return self.b
    
    def addToA(self, number):
        return self.a + number

In [103]:
s = Something(1)

In [104]:
s.returnA()

1

In [102]:
s.addToA(10)

11

In [105]:
s.returnB()

3

## Expressions

In [113]:
1 and 0

0

In [114]:
1 or 0

1

In [115]:
not 0

True

In [116]:
1 == True

True

In [117]:
1 >> 0

1

In [118]:
1 >= 0

True

In [119]:
1 != 0

True

In [147]:
2 > 1 > 0

True

In [120]:
1*3

3

In [121]:
1/3

0.3333333333333333

In [122]:
1//3

0

In [146]:
15//10

1

In [154]:
-15//10

-2

In [123]:
1%3

1

In [126]:
2**2

4

In [148]:
x = 1

In [149]:
y = 2.0

Autoconversion to the most complex type

In [150]:
type(x)

int

In [151]:
type(y)

float

In [152]:
z = x + y

In [153]:
type(z)

float

## Bitwise expressions

In [135]:
b1 = 0b0001

In [136]:
b2 = 0b0011

In [141]:
bin(b1 & b2)

'0b1'

In [142]:
bin(b1 | b2)

'0b11'

In [143]:
bin(b1 ^ b2)

'0b10'

In [144]:
bin(b1<<1)

'0b10'

In [145]:
bin(b1>>1)

'0b0'

In [156]:
bin(~b2)

'-0b100'

In [158]:
(~b2).bit_length()

3

## Dec, Hex, Bin and Oct representation

In [159]:
bin(12)

'0b1100'

In [160]:
hex(12)

'0xc'

In [161]:
oct(12)

'0o14'

In [353]:
'{:b}'.format(12)

'1100'

In [354]:
'{:8b}'.format(12)

'    1100'

In [355]:
'{:08b}'.format(12)

'00001100'

In [167]:
'{:x}'.format(12)

'c'

In [168]:
'{:o}'.format(12)

'14'

In [169]:
'{:X}'.format(12)

'C'

In [252]:
int('1111',2)

15

In [253]:
int('F',16)

15

In [256]:
int('17',8)

15

## Fractions

In [170]:
from fractions import Fraction

In [171]:
x = Fraction(1,3)

In [172]:
y = Fraction(2,6)

In [173]:
x + y

Fraction(2, 3)

In [174]:
print(x+y)

2/3


## References

In [175]:
a = 1

In [178]:
b = a

In [179]:
a == b

True

In [180]:
a is b

True

In [181]:
b = 2

In [182]:
a == b

False

In [183]:
a is b

False

Do the same for lists ;)

In [184]:
l1 = [1,2,3]

In [186]:
l1 == l2

True

In [187]:
l2 is l1

True

In [188]:
l2[0] = 1111

In [189]:
l2 == l1

True

In [190]:
l2 is l1

True

In [191]:
l1

[1111, 2, 3]

## Strings

In [270]:
x = 'abcdefghijklmnoprstuvwyx'

In [271]:
y = '''multiline
string'''

In [272]:
x[0]

'a'

In [273]:
x[0:-1]

'abcdefghijklmnoprstuvwy'

In [274]:
x[:-1]

'abcdefghijklmnoprstuvwy'

In [275]:
x[::10]

'akv'

In [276]:
x[::-1]

'xywvutsrponmlkjihgfedcba'

In [277]:
x = 'abc\ndef\n'

In [278]:
x

'abc\ndef\n'

In [279]:
print(x)

abc
def



In [280]:
x = 'test\'test'

In [281]:
print(x)

test'test


In [282]:
x = 'abc'

In [283]:
len(x)

3

In [284]:
x.find('b')

1

In [285]:
x.split()

['abc']

In [286]:
x.upper()

'ABC'

In [287]:
x.endswith('c')

True

In [288]:
x.encode('utf-8')

b'abc'

In [293]:
'abcTESTefgTEST'.replace('TEST','XXX')

'abcXXXefgXXX'

In [294]:
'abcTESTefgTEST'.replace('TEST','XXX',1)

'abcXXXefgTEST'

In [289]:
'środa'.encode('utf-8')

b'\xc5\x9broda'

In [290]:
[c*2 for c in x]

['aa', 'bb', 'cc']

In [291]:
''.join([c*2 for c in x])

'aabbcc'

In [267]:
list('abc')

['a', 'b', 'c']

Characters conversion

In [250]:
ord('a')

97

In [251]:
chr(97)

'a'

## String formating

In [337]:
'This is %s to show how to format strings (%d)' % ('code', 1)

'This is code to show how to format strings (1)'

In [338]:
'This is {} to show how to format strings ({})'.format('code', 1)

'This is code to show how to format strings (%d)'

In [340]:
'This is {1} to show how to format strings ({0})'.format(1, 'code')

'This is code to show how to format strings (1)'

In [341]:
'This is {something1} to show how to format strings ({something2})'.format(something1='code', something2=1)

'This is code to show how to format strings (1)'

In [343]:
'This is {d[test1]} to show how to format strings ({d[test2]})'.format(d={'test1':'code', 'test2':1})

'This is code to show how to format strings (1)'

In [345]:
class test:
    var1 = 'code'
    var2 = 1

In [348]:
'This is {o.var1} tho show how to format strings ({o.var2})'.format(o=test())

'This is code tho show how to format strings (1)'

## Conditional statements

In [356]:
if True:
    print('yeeeeah')

yeeeeah


In [357]:
if False:
    print(':(')
else:
    print('yeaaaah')

yeaaaah


In [358]:
if False:
    print(':(')
elif 0:
    print(':(')
else:
    print('yeaaaah')

yeaaaah


In [360]:
if ((False and False) or (False and False)) and not (True and True):
    print(':(')
else:
    print(';)')

;)


## While loop

In [363]:
i = 0
while i<3:
    print(i)
    i+=1

0
1
2


In [2]:
i = 0
while True:
    print(i)
    i+=1
    if i>=3:
        break

0
1
2


## Multithreading

In [19]:
from threading import Thread
from datetime import datetime

def count_something(limit):
    for i in range(1,limit):
        i*i

start_time = datetime.now()
t1 = Thread(target=count_something, args=(50000000,))
t1.start()
t1.join()
end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

Duration: 0:00:04.412654


In [20]:
from threading import Thread
from datetime import datetime

def count_something(limit):
    for i in range(1,limit):
        i*i

start_time = datetime.now()
t1 = Thread(target=count_something, args=(25000000,))
t2 = Thread(target=count_something, args=(25000000,))
t1.start()
t2.start()
t1.join()
t2.join()
end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

Duration: 0:00:04.803934


4.41 (one thread) vs 4.80 (two threads) - bacause of GIL (https://realpython.com/python-gil/)